Web Inspector: support OffscreenCanvas for Canvas related operations

kkinnunen-apple opened the pull request:
#### 552a34376f9678e24c422e71dfdb3539bc9b096e
<pre>
Web Inspector: support OffscreenCanvas for Canvas related operations
<a href="https://bugs.webkit.org/show_bug.cgi?id=180833">https://bugs.webkit.org/show_bug.cgi?id=180833</a>
rdar://36059660

Reviewed by Devin Rousso.

Implement Inspector support for OffscreenCanvas
OffscreenCanvasRenderingContext2D for main page (not workers).
Later commits may add support for workers.

Fixes assertion failures when bringing up Inspector for pages
that use OffscreenCanvas.

* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d.html: Added.
* LayoutTests/inspector/canvas/context-attributes-expected.txt:
* LayoutTests/inspector/canvas/context-attributes.html:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-2d.html:
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-saves-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-saves.html: Added.
* LayoutTests/inspector/canvas/requestClientNodes-expected.txt:
* LayoutTests/inspector/canvas/requestClientNodes.html:
* LayoutTests/inspector/canvas/requestContent-2d-expected.txt:
* LayoutTests/inspector/canvas/requestContent-2d.html:
* LayoutTests/inspector/canvas/resolveContext-2d-expected.txt:
* LayoutTests/inspector/canvas/resolveContext-2d.html:
* LayoutTests/inspector/canvas/resources/create-context-utilities.js:

Add new tests for test runner .html files for tests that use recording-2d.js.
The current test driver logic is a bit too convoluted to run these tests in
the same .html runner as non-offscreen variants. Later commits can refactor
the tests if needed.

For other tests, add the tests in the respective -2d.html file.

(createDetachedCanvas):
(createCSSCanvas):
(createOffscreenCanvas):
(destroyCanvases):
* LayoutTests/inspector/canvas/resources/recording-2d.js:
(async load):
(ignoreException): Deleted.
(cancelActions): Deleted.
(performActions.executeFrameFunction): Deleted.
(performActions): Deleted.
(performConsoleActions): Deleted.
(performSavePreActions.saveAndSet): Deleted.
(performSavePreActions): Deleted.
(performSavePostActions): Deleted.
* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::getContext):
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::is2dBase const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::create):
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.idl:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::resolveContext const):
(WebCore::InspectorCanvas::canvasChanged):
(WebCore::buildObjectForCanvasContextAttributes):
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::releaseObjectForRecording):
(WebCore::InspectorCanvas::getContentAsDataURL):
(WebCore::InspectorCanvas::appendActionSnapshotIfNeeded):
(WebCore::InspectorCanvas::buildInitialState):
(WebCore::InspectorCanvas::getCanvasContentAsDataURL): Deleted.
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/inspector/InspectorCanvasCallTracer.cpp:
(WebCore::InspectorCanvasCallTracer::processArgument):
(WebCore::InspectorCanvasCallTracer::recordAction):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
(WebCore::InspectorCanvasCallTracer::recordAction):
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
(WebCore::InspectorCanvasAgent::enable):
(WebCore::InspectorCanvasAgent::requestContent):
(WebCore::InspectorCanvasAgent::didChangeCanvasMemory):
(WebCore::InspectorCanvasAgent::startRecording):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::canvasRenderingContext):
(WebCore::PageConsoleClient::screenshot):
(WebCore::snapshotCanvas): Deleted.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas.fromPayload):
(WI.Canvas.displayNameForContextType):
* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording.fromPayload):
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.createContext):
(WI.Recording.prototype.async _process):
(WI.Recording):
* Source/WebInspectorUI/UserInterface/Models/RecordingAction.js:
(WI.RecordingAction._prototypeForType):
(WI.RecordingAction.prototype.process):
(WI.RecordingAction.prototype.async swizzle):
* Source/WebInspectorUI/UserInterface/Models/RecordingState.js:
(WI.RecordingState.prototype.fromContext):
(WI.RecordingState.async swizzleInitialState):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/CanvasSidebarPanel.js:
(WI.CanvasSidebarPanel.prototype._canvasChanged):
(WI.CanvasSidebarPanel.prototype._updateRecordNavigationItem):
* Source/WebInspectorUI/UserInterface/Views/RecordingContentView.js:
(WI.RecordingContentView):
(WI.RecordingContentView.prototype.get navigationItems):
(WI.RecordingContentView.prototype.attached):
(WI.RecordingContentView.prototype._export):
(WI.RecordingContentView.prototype._updateExportButton):
(WI.RecordingContentView.prototype._handleExportNavigationItemClicked):
* Source/WebInspectorUI/UserInterface/Views/RecordingStateDetailsSidebarPanel.js:
(WI.RecordingStateDetailsSidebarPanel.prototype.inspect):
(WI.RecordingStateDetailsSidebarPanel.prototype.set action):

Canonical link: <a href="https://commits.webkit.org/262388@main">https://commits.webkit.org/262388@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3c51acbd174cbacbb40e7298b4f77deec935b32a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2043 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1120 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1236 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1353 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1278 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1260 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/1206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1163 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1918 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1193 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1152 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1138 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/1099 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1145 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1183 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2253 "run-api-tests-without-change (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/1226 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1193 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1093 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/1283 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/1130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1162 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/259 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/351 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1220 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/1286 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/255 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Canvas.json:
```diff
@@ -23,7 +23,7 @@
             "id": "ContextType",
             "type": "string",
-            "enum": ["canvas-2d", "bitmaprenderer", "webgl", "webgl2"],
+            "enum": ["canvas-2d", "offscreen-canvas-2d", "bitmaprenderer", "webgl", "webgl2"],
```
rather than differentiate a different `ContextType`, would it better to have a `ContextSource` (or `ContextCreator` or something) with `HTMLCanvasElement` and `OffscreenCanvas`?  that way the frontend can still operate basically the same (i.e. a 2d context is a 2d context anywhere it is (for the most part))

kkinnunen-apple replied:
Not sure about this one yet, I'm not that familiar with frontend/backend of inspector. Maybe a future refactoring?

dcrousso replied:
i suppose since there's a different API surface it would be better to do it this way.  it would also make it easier for `Recording` since we wouldn't need another thing there (e.g. `Recording.Source` or `Recording.Creator` or etc.)

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Recording.js:
```diff
@@ -153,6 +156,8 @@ WI.Recording = class Recording extends WI.Object
         case Recording.Type.Canvas2D:
             return WI.UIString("2D");
+        case Recording.Type.OffscreenCanvas2D:
+                return WI.UIString("Offscreen 2D");
```
NIT: I think this would be better as `"2D (Offscreen)"`

kkinnunen-apple replied:
renamed.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/Recording.js:
```diff
@@ -555,7 +573,7 @@ WI.Recording = class Recording extends WI.Object
     toHTML()
     {
-        console.assert(this._type === WI.Recording.Type.Canvas2D);
+        console.assert(this._type === WI.Recording.Type.Canvas2D || this._type === WI.Recording.Type.OffscreenCanvas2D);
```
maybe we should make a
```
static is2D(contextType)
{
    switch (contextType)
    case WI.Recording.Type.Canvas2D:
    case WI.Recording.Type.OffscreenCanvas2D:
        return true;
    }
    return false;
}
```
and use that everywhere instead of having to repeat it

kkinnunen-apple replied:
done.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/RecordingAction.js:
```diff
@@ -727,6 +728,8 @@ WI.RecordingAction._visualNames = {
     ]),
 };
 
+WI.RecordingAction._visualNames[WI.Recording.Type.OffscreenCanvas2D] = WI.RecordingAction._visualNames[WI.Recording.Type.Canvas2D];
```
this actually isnt entirely true as there are methods on `CanvasRenderingContext` that are not on `OffscreenCanvasRenderingContext`

kkinnunen-apple replied:
Added only the methods that are exposed from OffscreenCanvasRenderingContext.idl.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/RecordingContentView.js:
```diff
@@ -31,7 +31,7 @@ WI.RecordingContentView = class RecordingContentView extends WI.ContentView
         super(representedObject);
 
-        let isCanvas2D = this.representedObject.type === WI.Recording.Type.Canvas2D;
+        let isCanvas2D = this.representedObject.isCanvas2D();
```
where is this defined?

kkinnunen-apple replied:
Added as Canvas.js Canvas.is2D()
(erroneous last-minute removal)

dcrousso commented on LayoutTests/inspector/canvas/resources/recording-2d.js:
```diff
@@ -84,6 +88,8 @@ function performActions() {
         },
         () => {
+            if (ctx instanceof OffscreenCanvasRenderingContext2D)
+                return;
```
NIT: instead of early returning you could use optional chaining to only call the method if it exists
```
ctx.clearShadow?.();
```

kkinnunen-apple replied:
done.

dcrousso reviewed (approved):
awesome! 🤩

are you planning to do worker stuff in/using #2873?

kkinnunen-apple committed:
Web Inspector: support OffscreenCanvas for Canvas related operations
https://bugs.webkit.org/show_bug.cgi?id=180833
rdar://36059660

Reviewed by Devin Rousso.

Implement Inspector support for OffscreenCanvas
OffscreenCanvasRenderingContext2D for main page (not workers).
Later commits may add support for workers.

Fixes assertion failures when bringing up Inspector for pages
that use OffscreenCanvas.

* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d-expected.txt: Added.
* LayoutTests/inspector/canvas/console-record-offscreen-canvas-2d.html: Added.
* LayoutTests/inspector/canvas/context-attributes-expected.txt:
* LayoutTests/inspector/canvas/context-attributes.html:
* LayoutTests/inspector/canvas/create-context-2d-expected.txt:
* LayoutTests/inspector/canvas/create-context-2d.html:
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-frameCount-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-frameCount.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-full-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-full.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-memoryLimit-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-memoryLimit.html: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-saves-expected.txt: Added.
* LayoutTests/inspector/canvas/recording-offscreen-canvas-2d-saves.html: Added.
* LayoutTests/inspector/canvas/requestClientNodes-expected.txt:
* LayoutTests/inspector/canvas/requestClientNodes.html:
* LayoutTests/inspector/canvas/requestContent-2d-expected.txt:
* LayoutTests/inspector/canvas/requestContent-2d.html:
* LayoutTests/inspector/canvas/resolveContext-2d-expected.txt:
* LayoutTests/inspector/canvas/resolveContext-2d.html:
* LayoutTests/inspector/canvas/resources/create-context-utilities.js:

Add new tests for test runner .html files for tests that use recording-2d.js.
The current test driver logic is a bit too convoluted to run these tests in
the same .html runner as non-offscreen variants. Later commits can refactor
the tests if needed.

For other tests, add the tests in the respective -2d.html file.

(createDetachedCanvas):
(createCSSCanvas):
(createOffscreenCanvas):
(destroyCanvases):
* LayoutTests/inspector/canvas/resources/recording-2d.js:
(async load):
(ignoreException): Deleted.
(cancelActions): Deleted.
(performActions.executeFrameFunction): Deleted.
(performActions): Deleted.
(performConsoleActions): Deleted.
(performSavePreActions.saveAndSet): Deleted.
(performSavePreActions): Deleted.
(performSavePostActions): Deleted.
* Source/JavaScriptCore/inspector/protocol/Canvas.json:
* Source/JavaScriptCore/inspector/protocol/Recording.json:
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::getContext):
* Source/WebCore/html/OffscreenCanvas.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::is2dBase const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.cpp:
(WebCore::OffscreenCanvasRenderingContext2D::create):
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.h:
* Source/WebCore/html/canvas/OffscreenCanvasRenderingContext2D.idl:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::resolveContext const):
(WebCore::InspectorCanvas::canvasChanged):
(WebCore::buildObjectForCanvasContextAttributes):
(WebCore::InspectorCanvas::buildObjectForCanvas):
(WebCore::InspectorCanvas::releaseObjectForRecording):
(WebCore::InspectorCanvas::getContentAsDataURL):
(WebCore::InspectorCanvas::appendActionSnapshotIfNeeded):
(WebCore::InspectorCanvas::buildInitialState):
(WebCore::InspectorCanvas::getCanvasContentAsDataURL): Deleted.
* Source/WebCore/inspector/InspectorCanvas.h:
* Source/WebCore/inspector/InspectorCanvasCallTracer.cpp:
(WebCore::InspectorCanvasCallTracer::processArgument):
(WebCore::InspectorCanvasCallTracer::recordAction):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
(WebCore::InspectorCanvasCallTracer::recordAction):
* Source/WebCore/inspector/agents/InspectorCanvasAgent.cpp:
(WebCore::InspectorCanvasAgent::enable):
(WebCore::InspectorCanvasAgent::requestContent):
(WebCore::InspectorCanvasAgent::didChangeCanvasMemory):
(WebCore::InspectorCanvasAgent::startRecording):
* Source/WebCore/page/PageConsoleClient.cpp:
(WebCore::canvasRenderingContext):
(WebCore::PageConsoleClient::screenshot):
(WebCore::snapshotCanvas): Deleted.
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Models/Canvas.js:
(WI.Canvas.fromPayload):
(WI.Canvas.displayNameForContextType):
* Source/WebInspectorUI/UserInterface/Models/Recording.js:
(WI.Recording.fromPayload):
(WI.Recording.displayNameForRecordingType):
(WI.Recording.prototype.createContext):
(WI.Recording.prototype.async _process):
(WI.Recording):
* Source/WebInspectorUI/UserInterface/Models/RecordingAction.js:
(WI.RecordingAction._prototypeForType):
(WI.RecordingAction.prototype.process):
(WI.RecordingAction.prototype.async swizzle):
* Source/WebInspectorUI/UserInterface/Models/RecordingState.js:
(WI.RecordingState.prototype.fromContext):
(WI.RecordingState.async swizzleInitialState):
* Source/WebInspectorUI/UserInterface/Views/CanvasContentView.js:
(WI.CanvasContentView.prototype.initialLayout):
* Source/WebInspectorUI/UserInterface/Views/CanvasSidebarPanel.js:
(WI.CanvasSidebarPanel.prototype._canvasChanged):
(WI.CanvasSidebarPanel.prototype._updateRecordNavigationItem):
* Source/WebInspectorUI/UserInterface/Views/RecordingContentView.js:
(WI.RecordingContentView):
(WI.RecordingContentView.prototype.get navigationItems):
(WI.RecordingContentView.prototype.attached):
(WI.RecordingContentView.prototype._export):
(WI.RecordingContentView.prototype._updateExportButton):
(WI.RecordingContentView.prototype._handleExportNavigationItemClicked):
* Source/WebInspectorUI/UserInterface/Views/RecordingStateDetailsSidebarPanel.js:
(WI.RecordingStateDetailsSidebarPanel.prototype.inspect):
(WI.RecordingStateDetailsSidebarPanel.prototype.set action):

Canonical link: https://commits.webkit.org/262388@main
